odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2385 @@
+
+        >>> tset = odl.tensor_set((2, 3), 'uint8')
+        >>> x = tset.element([[0, 0, 0],
+        ...                   [1, 1, 1]])
+        >>> maxval = 255  # maximum signed 8-bit unsigned int
+        >>> x[0, :] = maxval + 1
+        >>> x
+        tensor_set((2, 3), 'uint8').element(
+            [[0, 0, 0],
+             [1, 1, 1]]
+        )
+        """
+        if isinstance(values, NumpyGeneralizedTensor):
+            self.data[indices] = values.data
+        else:
+            self.data[indices] = values
+
+    @property
+    def ufuncs(self):
+        """`NumpyTensorSetUfuncs`, access to Numpy style ufuncs.
+
+        Notes
+        -----
+        These ufuncs are optimized for use with `NumpyGeneralizedTensor`'s
+        and incur practically no overhead.
+
+        Examples
+        --------
+        >>> space = odl.rn((2, 3))
+        >>> x = space.element([[1, -2, 3],
+        ...                    [4, -5, 6]])
+        >>> x.ufuncs.absolute()
+        rn((2, 3)).element(
+            [[1.0, 2.0, 3.0],
+             [4.0, 5.0, 6.0]]
+        )
+
+        Broadcasting and array-like input is supported, too:
+
+        >>> x.ufuncs.add(3)
+        rn((2, 3)).element(
+            [[4.0, 1.0, 6.0],
+             [7.0, -2.0, 9.0]]
+        )
+        >>> x.ufuncs.subtract([[0, 0, 1],
+        ...                   [1, 0, 0]])
+        rn((2, 3)).element(
+            [[1.0, -2.0, 2.0],
+             [3.0, -5.0, 6.0]]
+        )
+
+        There is also support for various reductions (sum, prod, min,
+        max):
+
+        >>> x.ufuncs.sum()
+        7.0
+
+        The ``out`` parameter can be used for result storage:
+
+        >>> y = space.one()
+        >>> out = space.element()
+        >>> result = x.ufuncs.add(y, out=out)
+        >>> result
+        rn((2, 3)).element(
+            [[2.0, -1.0, 4.0],
+             [5.0, -4.0, 7.0]]
+        )
+        >>> result is out
+        True
+        """
+        return NumpyTensorSetUfuncs(self)
+
+
+def _blas_is_applicable(*args):
+    """Whether BLAS routines can be applied or not.
+
+    BLAS routines are available for single and double precision
+    float or complex data only. If the arrays are non-contiguous,
+    BLAS methods are usually slower, and array-writing routines do
+    not work at all. Hence, only contiguous arrays are allowed.
+
+    Parameters
+    ----------
+    x1,...,xN : `NumpyGeneralizedTensor`
+        The tensors to be tested for BLAS conformity.
+    """
+    if any(x.dtype != args[0].dtype for x in args[1:]):
+        return False
+    if any(x.dtype not in _BLAS_DTYPES for x in args):
+        return False
+    if not (all(x.flags.f_contiguous for x in args) or
+            all(x.flags.c_contiguous for x in args)):
+        return False
+    return True
+
+
+def _lincomb(a, x1, b, x2, out, dtype):
+    """Raw linear combination depending on data type."""
+    # Shortcut for small problems
+    # if x1.size < 100:  # small array optimization
```

Comment:
This is only temporarily commented out for testing. Before, the shortcut was hiding the fact that the implementation was completely broken since `out` was never modified.

Actually, I'm pretty sure the fallbacks won't work as they are now since they should work directly on the non-flattened arrays. If those are discontiguous, no flattening will preserve views, and then `out` won't be modified.
Obviously, this case is also lacking a test.